src/main/java/com/commercetools/paymenttoorderprocessor/jobs/actions/MessageReader.java

```diff
@@ -43,51 +44,72 @@
+    final int PAGEOVERLAP = 5;
 
+    /**
+     * @return the oldest unprocessed message from the queue, or <b>null</b> if no new messages to process in CTP.
+     */
     @Override
+    @Nullable
     public PaymentTransactionStateChangedMessage read() {
         LOG.debug("wasInitialQueried: {}", wasInitialQueried);
-        if (isQueryNeeded()) {
-            getUnprocessedMessagesFromPlatform();
+
+        while (isQueryNeeded()) {
+            fetchUnprocessedMessagesFromPlatform();
         }
-        return getMessageFromList();
+
+        return getUnprocessedMessageFromQueue();
     }
 
-    private PaymentTransactionStateChangedMessage getMessageFromList() {
-        if (messages.isEmpty()) {
-            return null;
-        } else {
-            timeStampManager.setActualProcessedMessageTimeStamp(messages.get(0).getLastModifiedAt());
-            return messages.remove(0);
-        }
+    /**
+     * @return oldest unprocessed message from the queue if exists, or <b>null</b>
+     */
+    @Nullable
+    private PaymentTransactionStateChangedMessage getUnprocessedMessageFromQueue() {
+        return unprocessedMessagesQueue.poll();
     }
 
+    /**
+     * A first or next CTP query is needed if we don't have unprocessed messages in the queue.
+     *
+     * @return <b>true</b> if messages queue never fetched or {@code unprocessedMessagesQueue} is empty and CTP still
+     * has items to fetch (query next page).
+     */
     private boolean isQueryNeeded() {
-        return !wasInitialQueried || (messages.isEmpty() && total > offset);
+        return !wasInitialQueried || (unprocessedMessagesQueue.isEmpty() && total > offset);
     }
 
-    private void getUnprocessedMessagesFromPlatform() {
-        final List<Message> result = queryPlatform();
-        messages = result.stream()
+    /**
+     * Fetch messages from the platform and put all unprocessed messages to {@code unprocessedMessagesQueue}.
+     * Also, update {@link TimeStampManager#setActualProcessedMessageTimeStamp(java.time.ZonedDateTime)
+     * actualProcessedMessageTimeStamp} for processed messages.
+     */
+    private void fetchUnprocessedMessagesFromPlatform() {
+        final PagedQueryResult<Message> result = queryPlatform();
+        result.getResults().stream()
                 .map(message -> message.as(PaymentTransactionStateChangedMessage.class))
-                .filter(message -> messageProcessedManager.isMessageUnprocessed(message))
-                .collect(Collectors.toList());
-
-        LOG.info("total {} messages, {} on current page, {} of them are are unprocessed", total, result.size(), messages.size());
+                .forEach(message -> {
+                    if (messageProcessedManager.isMessageUnprocessed(message)) {
+                        unprocessedMessagesQueue.add(message);
+                    } else {
+                        timeStampManager.setActualProcessedMessageTimeStamp(message.getLastModifiedAt());
```

Comment:
Could it be that you might skip some unprocessed messages? When you are iterating over the messages, can they be in this order: `[unprocessedMessage1, processedMessage, unprocessedMessage2]`. Then unprocessedMessage1 fails, you save `processedMessage` timestamp, and on the next round, you don't process unprocessedMessage1 anymore.